# First Step - Explore the data

In [2]:
%load_ext sql

In [4]:
%sql postgresql://ziv:ziv@localhost:5433/gis

'Connected: ziv@gis'

In [3]:
import pandas as pd
import geopandas as gpd
import psycopg2
import subprocess
import os

## Trips - csv

the follwoing dataset describe texi trips in nyc city

Below is heatmap of the pickup coordinates, the red area show a bigger cluster of points.

![Trips pickup heatmap](img/trips_pickup_heat.png)

In [49]:
trips_df = pd.read_csv('data/trips.csv')

In [50]:
trips_df.head()

id  vendor_id      pickup_datetime     dropoff_datetime  \
0  id2875421          2  2016-03-14 17:24:55  2016-03-14 17:32:30   
1  id2377394          1  2016-06-12 00:43:35  2016-06-12 00:54:38   
2  id3858529          2  2016-01-19 11:35:24  2016-01-19 12:10:48   
3  id3504673          2  2016-04-06 19:32:31  2016-04-06 19:39:40   
4  id2181028          2  2016-03-26 13:30:55  2016-03-26 13:38:10   

   passenger_count  pickup_longitude  pickup_latitude  dropoff_longitude  \
0                1        -73.982155        40.767937         -73.964630   
1                1        -73.980415        40.738564         -73.999481   
2                1        -73.979027        40.763939         -74.005333   
3                1        -74.010040        40.719971         -74.012268   
4                1        -73.973053        40.793209         -73.972923   

   dropoff_latitude store_and_fwd_flag  trip_duration  
0         40.765602                  N            455  
1         40.731152                  N            663  
2         40.710087                  N           2124  
3         40.706718                  N            429  
4         40.782520                  N            435

In [51]:
trips_df.tail()

id  vendor_id      pickup_datetime     dropoff_datetime  \
1458639  id2376096          2  2016-04-08 13:31:04  2016-04-08 13:44:02   
1458640  id1049543          1  2016-01-10 07:35:15  2016-01-10 07:46:10   
1458641  id2304944          2  2016-04-22 06:57:41  2016-04-22 07:10:25   
1458642  id2714485          1  2016-01-05 15:56:26  2016-01-05 16:02:39   
1458643  id1209952          1  2016-04-05 14:44:25  2016-04-05 14:47:43   

         passenger_count  pickup_longitude  pickup_latitude  \
1458639                4        -73.982201        40.745522   
1458640                1        -74.000946        40.747379   
1458641                1        -73.959129        40.768799   
1458642                1        -73.982079        40.749062   
1458643                1        -73.979538        40.781750   

         dropoff_longitude  dropoff_latitude store_and_fwd_flag  trip_duration  
1458639         -73.994911         40.740170                  N            778  
1458640         -73.970184         40.796547                  N            655  
1458641         -74.004433         40.707371                  N            764  
1458642         -73.974632         40.757107                  N            373  
1458643         -73.972809         40.790585                  N            198

In [52]:
trips_df.count()

id                    1458644
vendor_id             1458644
pickup_datetime       1458644
dropoff_datetime      1458644
passenger_count       1458644
pickup_longitude      1458644
pickup_latitude       1458644
dropoff_longitude     1458644
dropoff_latitude      1458644
store_and_fwd_flag    1458644
trip_duration         1458644
dtype: int64

In [54]:
trips_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1458644 entries, 0 to 1458643
Data columns (total 11 columns):
 #   Column              Non-Null Count    Dtype  
---  ------              --------------    -----  
 0   id                  1458644 non-null  object 
 1   vendor_id           1458644 non-null  int64  
 2   pickup_datetime     1458644 non-null  object 
 3   dropoff_datetime    1458644 non-null  object 
 4   passenger_count     1458644 non-null  int64  
 5   pickup_longitude    1458644 non-null  float64
 6   pickup_latitude     1458644 non-null  float64
 7   dropoff_longitude   1458644 non-null  float64
 8   dropoff_latitude    1458644 non-null  float64
 9   store_and_fwd_flag  1458644 non-null  object 
 10  trip_duration       1458644 non-null  int64  
dtypes: float64(4), int64(3), object(4)
memory usage: 122.4+ MB


In [78]:
print(f'Are there any NaN in the dataset? {trips_df.isnull().values.any()}, how many? {trips_df.isnull().sum().sum()}')

Are there any NaN in the dataset? False, how many? 0


In this dataset most of the data is numeric data type. the coordinates are the important fields for the data analyst in this case, and will need to be counted per polygon in the grid. 
The pickup/dropoff time also may be needed for analsys. 
There isn't any NaN value in the dataset, and we can load it as it, but for future dataset a qa logic maybe be needed

# Census dataset - csv

The census tract is a kind of fact table, with almost only metric columns.

In [5]:
census_df = pd.read_csv('../data/census_tracts.csv')

In [6]:
census_df.head(5)

CensusTract County Borough  TotalPop   Men  Women  Hispanic  White  Black  \
0  36005000100  Bronx   Bronx      7703  7133    570      29.9    6.1   60.9   
1  36005000200  Bronx   Bronx      5403  2659   2744      75.8    2.3   16.0   
2  36005000400  Bronx   Bronx      5915  2896   3019      62.7    3.6   30.7   
3  36005001600  Bronx   Bronx      5879  2558   3321      65.1    1.6   32.4   
4  36005001900  Bronx   Bronx      2591  1206   1385      55.4    9.0   29.0   

   Native  ...  Walk  OtherTransp  WorkAtHome  MeanCommute  Employed  \
0     0.2  ...   NaN          NaN         NaN          NaN         0   
1     0.0  ...   2.9          0.0         0.0         43.0      2308   
2     0.0  ...   1.4          0.5         2.1         45.0      2675   
3     0.0  ...   8.6          1.6         1.7         38.8      2120   
4     0.0  ...   3.0          2.4         6.2         45.4      1083   

   PrivateWork  PublicWork  SelfEmployed  FamilyWork  Unemployment  
0          NaN         NaN           NaN         NaN           NaN  
1         80.8        16.2           2.9         0.0           7.7  
2         71.7        25.3           2.5         0.6           9.5  
3         75.0        21.3           3.8         0.0           8.7  
4         76.8        15.5           7.7         0.0          19.2  

[5 rows x 36 columns]

In [7]:
census_df.count()

CensusTract        2167
County             2167
Borough            2167
TotalPop           2167
Men                2167
Women              2167
Hispanic           2128
White              2128
Black              2128
Native             2128
Asian              2128
Citizen            2167
Income             2101
IncomeErr          2101
IncomePerCap       2121
IncomePerCapErr    2121
Poverty            2125
ChildPoverty       2107
Professional       2124
Service            2124
Office             2124
Construction       2124
Production         2124
Drive              2124
Carpool            2124
Transit            2124
Walk               2124
OtherTransp        2124
WorkAtHome         2124
MeanCommute        2106
Employed           2167
PrivateWork        2124
PublicWork         2124
SelfEmployed       2124
FamilyWork         2124
Unemployment       2125
dtype: int64

In [8]:
census_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2167 entries, 0 to 2166
Data columns (total 36 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CensusTract      2167 non-null   int64  
 1   County           2167 non-null   object 
 2   Borough          2167 non-null   object 
 3   TotalPop         2167 non-null   int64  
 4   Men              2167 non-null   int64  
 5   Women            2167 non-null   int64  
 6   Hispanic         2128 non-null   float64
 7   White            2128 non-null   float64
 8   Black            2128 non-null   float64
 9   Native           2128 non-null   float64
 10  Asian            2128 non-null   float64
 11  Citizen          2167 non-null   int64  
 12  Income           2101 non-null   float64
 13  IncomeErr        2101 non-null   float64
 14  IncomePerCap     2121 non-null   float64
 15  IncomePerCapErr  2121 non-null   float64
 16  Poverty          2125 non-null   float64
 17  ChildPoverty  

The dataset include few NaN in varius rows. The diffrence between the the total diffrence of NaN isn't large between the fields, so we can remove them.

In [9]:
print(f'Are there any NaN in the dataset? {census_df.isnull().values.any()}, how many? {census_df.isnull().sum().sum()}')

Are there any NaN in the dataset? True, how many? 1269


In [10]:
census_df_clean = census_df.dropna(how='any',axis=0) 

In [11]:
census_df_clean.head()

CensusTract County Borough  TotalPop   Men  Women  Hispanic  White  Black  \
1  36005000200  Bronx   Bronx      5403  2659   2744      75.8    2.3   16.0   
2  36005000400  Bronx   Bronx      5915  2896   3019      62.7    3.6   30.7   
3  36005001600  Bronx   Bronx      5879  2558   3321      65.1    1.6   32.4   
4  36005001900  Bronx   Bronx      2591  1206   1385      55.4    9.0   29.0   
5  36005002000  Bronx   Bronx      8516  3301   5215      61.1    1.6   31.1   

   Native  ...  Walk  OtherTransp  WorkAtHome  MeanCommute  Employed  \
1     0.0  ...   2.9          0.0         0.0         43.0      2308   
2     0.0  ...   1.4          0.5         2.1         45.0      2675   
3     0.0  ...   8.6          1.6         1.7         38.8      2120   
4     0.0  ...   3.0          2.4         6.2         45.4      1083   
5     0.3  ...   4.3          1.0         0.0         46.0      2508   

   PrivateWork  PublicWork  SelfEmployed  FamilyWork  Unemployment  
1         80.8        16.2           2.9         0.0           7.7  
2         71.7        25.3           2.5         0.6           9.5  
3         75.0        21.3           3.8         0.0           8.7  
4         76.8        15.5           7.7         0.0          19.2  
5         71.0        21.3           7.7         0.0          17.2  

[5 rows x 36 columns]

In [12]:
print(f'Are there any NaN in the dataset? {census_df_clean.isnull().values.any()}, how many? {census_df_clean.isnull().sum().sum()}')
census_df_clean.info()

Are there any NaN in the dataset? False, how many? 0
<class 'pandas.core.frame.DataFrame'>
Int64Index: 2095 entries, 1 to 2165
Data columns (total 36 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CensusTract      2095 non-null   int64  
 1   County           2095 non-null   object 
 2   Borough          2095 non-null   object 
 3   TotalPop         2095 non-null   int64  
 4   Men              2095 non-null   int64  
 5   Women            2095 non-null   int64  
 6   Hispanic         2095 non-null   float64
 7   White            2095 non-null   float64
 8   Black            2095 non-null   float64
 9   Native           2095 non-null   float64
 10  Asian            2095 non-null   float64
 11  Citizen          2095 non-null   int64  
 12  Income           2095 non-null   float64
 13  IncomeErr        2095 non-null   float64
 14  IncomePerCap     2095 non-null   float64
 15  IncomePerCapErr  2095 non-null   float64
 16  Poverty

In [13]:
census_df_clean.describe()

CensusTract      TotalPop           Men         Women     Hispanic  \
count  2.095000e+03   2095.000000   2095.000000   2095.000000  2095.000000   
mean   3.605464e+10   4008.149403   1907.002387   2101.147017    26.703771   
std    2.589809e+07   2167.464345   1023.925010   1176.607752    22.544082   
min    3.600500e+10    120.000000     58.000000     62.000000     0.000000   
25%    3.604703e+10   2463.500000   1176.500000   1274.000000     9.100000   
50%    3.604712e+10   3626.000000   1728.000000   1873.000000    18.500000   
75%    3.608103e+10   5010.500000   2392.500000   2603.000000    39.850000   
max    3.608503e+10  28926.000000  13460.000000  15466.000000    95.600000   

             White        Black       Native        Asian       Citizen  ...  \
count  2095.000000  2095.000000  2095.000000  2095.000000   2095.000000  ...   
mean     32.831647    23.948878     0.195847    13.585585   2509.210024  ...   
std      30.251354    29.463121     0.673646    16.470684   1500.855741  ...   
min       0.000000     0.000000     0.000000     0.000000     68.000000  ...   
25%       4.050000     1.500000     0.000000     2.200000   1513.000000  ...   
50%      22.700000     7.800000     0.000000     6.800000   2166.000000  ...   
75%      59.800000    38.950000     0.000000    19.250000   3009.500000  ...   
max     100.000000    97.800000    11.300000    89.800000  22905.000000  ...   

              Walk  OtherTransp   WorkAtHome  MeanCommute      Employed  \
count  2095.000000  2095.000000  2095.000000  2095.000000   2095.000000   
mean      8.937184     2.281384     3.609928    40.900286   1871.298329   
std       9.050245     2.566558     3.173333     7.021246   1152.469508   
min       0.000000     0.000000     0.000000    17.100000     57.000000   
25%       3.300000     0.500000     1.300000    37.300000   1102.000000   
50%       6.500000     1.500000     2.800000    41.400000   1622.000000   
75%      10.700000     3.200000     5.100000    45.400000   2298.500000   
max      60.100000    18.900000    25.200000    70.500000  12780.000000   

       PrivateWork   PublicWork  SelfEmployed   FamilyWork  Unemployment  
count  2095.000000  2095.000000   2095.000000  2095.000000   2095.000000  
mean     79.467494    14.258950      6.152983     0.120286      9.773604  
std       6.593343     7.054438      3.521428     0.394014      5.258753  
min      48.200000     0.000000      0.000000     0.000000      0.000000  
25%      75.000000     8.900000      3.600000     0.000000      6.000000  
50%      79.800000    13.400000      5.700000     0.000000      8.900000  
75%      84.300000    18.700000      8.200000     0.000000     12.600000  
max     100.000000    39.900000     23.300000     3.900000     62.700000  

[8 rows x 34 columns]

it seems that there a large std in the men and woman rows, and min max show it. to make sure there isn't a problem with the data let's choose the row. the mean of women and men is similar, 2101 women and 1907 men, so the total pop distribution is  good.

In [14]:
census_df_clean.loc[census_df_clean['Women'] == 62]

CensusTract    County    Borough  TotalPop  Men  Women  Hispanic  White  \
1195  36061009600  New York  Manhattan       149   87     62       4.7   55.0   
1764  36081056100    Queens     Queens       120   58     62      12.5   23.3   

      Black  Native  ...  Walk  OtherTransp  WorkAtHome  MeanCommute  \
1195    0.0     8.1  ...  58.6          0.0         7.8         25.7   
1764   37.5     0.0  ...   6.7          0.0         0.0         35.3   

      Employed  PrivateWork  PublicWork  SelfEmployed  FamilyWork  \
1195       116        100.0         0.0           0.0         0.0   
1764        64         71.9        21.9           6.3         0.0   

      Unemployment  
1195           4.1  
1764           0.0  

[2 rows x 36 columns]

In [15]:
census_df_clean.loc[census_df_clean['Women'] == 15466]

CensusTract County Borough  TotalPop    Men  Women  Hispanic  White  \
334  36005046201  Bronx   Bronx     28926  13460  15466      31.9    9.5   

     Black  Native  ...  Walk  OtherTransp  WorkAtHome  MeanCommute  Employed  \
334   56.1     0.1  ...   7.0          1.1         1.2         54.9     12780   

     PrivateWork  PublicWork  SelfEmployed  FamilyWork  Unemployment  
334         66.9        31.7           1.4         0.0          11.8  

[1 rows x 36 columns]

# Census block location dataset - csv

The following dataset include both NJ and NY states, while the other datasets include only NY. Also there are not any NaN.

In [17]:
census_blocks_df = pd.read_csv('../data/census_block_loc.csv')

In [18]:
census_blocks_df.head(10)

Latitude  Longitude        BlockCode     County State
0     40.48 -74.280000  340230076002012  Middlesex    NJ
1     40.48 -74.276834  340230076005000  Middlesex    NJ
2     40.48 -74.273668  340230076003018  Middlesex    NJ
3     40.48 -74.270503  340230076003004  Middlesex    NJ
4     40.48 -74.267337  340230074021000  Middlesex    NJ
5     40.48 -74.264171  340230074021000  Middlesex    NJ
6     40.48 -74.261005  340230074021000  Middlesex    NJ
7     40.48 -74.257839  340230074021000  Middlesex    NJ
8     40.48 -74.254673  340230074025000  Middlesex    NJ
9     40.48 -74.251508  340230074025000  Middlesex    NJ

In [19]:
census_blocks_df.count()

Latitude     38396
Longitude    38396
BlockCode    38396
County       38396
State        38396
dtype: int64

In [20]:
print(f'Are there any NaN in the dataset? {census_blocks_df.isnull().values.any()}, how many? {census_blocks_df.isnull().sum().sum()}')
census_blocks_df.info()

Are there any NaN in the dataset? False, how many? 0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38396 entries, 0 to 38395
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Latitude   38396 non-null  float64
 1   Longitude  38396 non-null  float64
 2   BlockCode  38396 non-null  int64  
 3   County     38396 non-null  object 
 4   State      38396 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 1.5+ MB


For any gis analsys a join is in need.

as shown here: https://www.kaggle.com/muonneutrino/mapping-new-york-city-census-data

> To merge our census data with the location data, we'll first need to identify the census tracts associated with our >locations. Here, I'll just use the data points in the census block file, but in general, we'll want to use the census block  > data as a lookup table.
> 
>To do this, we take our coordinates (i.e. the census block data) and perform a left join with the tract values as keys. > Since the census data is only for New York City itself, we'll first filter out everything outside the city.

In [21]:
blocks_df = census_blocks_df[census_blocks_df.County.isin(['Bronx','Kings','New York','Queens','Richmond'])]
blocks_df['Tract'] = blocks_df.BlockCode // 10000
blocks_df.head()

<ipython-input-21-2fbc11303c4b>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  blocks_df['Tract'] = blocks_df.BlockCode // 10000


Latitude  Longitude        BlockCode    County State        Tract
15     40.48 -74.232513  360859901000011  Richmond    NY  36085990100
16     40.48 -74.229347  360859901000011  Richmond    NY  36085990100
17     40.48 -74.226181  360859901000011  Richmond    NY  36085990100
18     40.48 -74.223015  360859901000011  Richmond    NY  36085990100
19     40.48 -74.219849  360859901000011  Richmond    NY  36085990100

In [22]:
blocks_df.count()

Latitude     18053
Longitude    18053
BlockCode    18053
County       18053
State        18053
Tract        18053
dtype: int64

In [23]:
blocks_merge = blocks_df.merge(census_df_clean,left_on='Tract', right_on='CensusTract')
blocks_merge.head(15)

Latitude  Longitude        BlockCode  County_x State        Tract  \
0   40.493568 -74.251508  360850244020002  Richmond    NY  36085024402   
1   40.493568 -74.248342  360850244020002  Richmond    NY  36085024402   
2   40.493568 -74.245176  360850244020002  Richmond    NY  36085024402   
3   40.493568 -74.242010  360850244020002  Richmond    NY  36085024402   
4   40.493568 -74.238844  360850244020002  Richmond    NY  36085024402   
5   40.495829 -74.254673  360850244020003  Richmond    NY  36085024402   
6   40.495829 -74.251508  360850244020003  Richmond    NY  36085024402   
7   40.495829 -74.248342  360850244020002  Richmond    NY  36085024402   
8   40.495829 -74.245176  360850244020002  Richmond    NY  36085024402   
9   40.495829 -74.242010  360850244020002  Richmond    NY  36085024402   
10  40.495829 -74.238844  360850244020002  Richmond    NY  36085024402   
11  40.495829 -74.235678  360850244020002  Richmond    NY  36085024402   
12  40.495829 -74.232513  360850244020001  Richmond    NY  36085024402   
13  40.495829 -74.229347  360850244020001  Richmond    NY  36085024402   
14  40.495829 -74.226181  360850244020001  Richmond    NY  36085024402   

    CensusTract  County_y        Borough  TotalPop  ...  Walk  OtherTransp  \
0   36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   
1   36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   
2   36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   
3   36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   
4   36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   
5   36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   
6   36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   
7   36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   
8   36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   
9   36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   
10  36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   
11  36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   
12  36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   
13  36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   
14  36085024402  Richmond  Staten Island      4241  ...   1.1          0.6   

    WorkAtHome  MeanCommute  Employed  PrivateWork  PublicWork  SelfEmployed  \
0          4.0         44.3      2046         75.2        21.2           3.6   
1          4.0         44.3      2046         75.2        21.2           3.6   
2          4.0         44.3      2046         75.2        21.2           3.6   
3          4.0         44.3      2046         75.2        21.2           3.6   
4          4.0         44.3      2046         75.2        21.2           3.6   
5          4.0         44.3      2046         75.2        21.2           3.6   
6          4.0         44.3      2046         75.2        21.2           3.6   
7          4.0         44.3      2046         75.2        21.2           3.6   
8          4.0         44.3      2046         75.2        21.2           3.6   
9          4.0         44.3      2046         75.2        21.2           3.6   
10         4.0         44.3      2046         75.2        21.2           3.6   
11         4.0         44.3      2046         75.2        21.2           3.6   
12         4.0         44.3      2046         75.2        21.2           3.6   
13         4.0         44.3      2046         75.2        21.2           3.6   
14         4.0         44.3      2046         75.2        21.2           3.6   

    FamilyWork  Unemployment  
0          0.0           8.3  
1          0.0           8.3  
2          0.0           8.3  
3          0.0           8.3  
4          0.0           8.3  
5          0.0           8.3  
6          0.0           8.3  
7          0.0           8.3  
8          

In [24]:
blocks_merge.count()

Latitude           11769
Longitude          11769
BlockCode          11769
County_x           11769
State              11769
Tract              11769
CensusTract        11769
County_y           11769
Borough            11769
TotalPop           11769
Men                11769
Women              11769
Hispanic           11769
White              11769
Black              11769
Native             11769
Asian              11769
Citizen            11769
Income             11769
IncomeErr          11769
IncomePerCap       11769
IncomePerCapErr    11769
Poverty            11769
ChildPoverty       11769
Professional       11769
Service            11769
Office             11769
Construction       11769
Production         11769
Drive              11769
Carpool            11769
Transit            11769
Walk               11769
OtherTransp        11769
WorkAtHome         11769
MeanCommute        11769
Employed           11769
PrivateWork        11769
PublicWork         11769
SelfEmployed       11769


In [25]:
blocks_merge.describe()

Latitude     Longitude     BlockCode         Tract   CensusTract  \
count  11769.000000  11769.000000  1.176900e+04  1.176900e+04  1.176900e+04   
mean      40.692854    -73.939690  3.606128e+14  3.606128e+10  3.606128e+10   
std        0.098142      0.133312  2.746956e+11  2.746956e+07  2.746956e+07   
min       40.493568    -74.257839  3.600500e+14  3.600500e+10  3.600500e+10   
25%       40.613417    -74.010905  3.604704e+14  3.604704e+10  3.604704e+10   
50%       40.685779    -73.922261  3.608102e+14  3.608102e+10  3.608102e+10   
75%       40.767186    -73.836784  3.608113e+14  3.608113e+10  3.608113e+10   
max       40.914171    -73.700653  3.608503e+14  3.608503e+10  3.608503e+10   

           TotalPop           Men         Women      Hispanic         White  \
count  11769.000000  11769.000000  11769.000000  11769.000000  11769.000000   
mean    4403.692752   2112.847906   2290.844847     22.955060     43.048296   
std     2608.938048   1235.078970   1406.007423     19.458575     31.765176   
min      120.000000     58.000000     62.000000      0.000000      0.000000   
25%     2693.000000   1318.000000   1385.000000      8.600000      9.300000   
50%     3994.000000   1881.000000   2037.000000     16.700000     45.800000   
75%     5693.000000   2749.000000   2933.000000     30.600000     70.200000   
max    28926.000000  13460.000000  15466.000000     95.600000    100.000000   

       ...          Walk   OtherTransp    WorkAtHome   MeanCommute  \
count  ...  11769.000000  11769.000000  11769.000000  11769.000000   
mean   ...      6.902490      1.941575      3.416883     41.404461   
std    ...      8.170206      2.576917      3.097461      6.900999   
min    ...      0.000000      0.000000      0.000000     17.100000   
25%    ...      1.900000      0.300000      1.200000     37.400000   
50%    ...      4.400000      1.100000      2.600000     41.500000   
75%    ...      8.500000      2.400000      4.600000     45.700000   
max    ...     60.100000     18.900000     25.200000     70.500000   

           Employed   PrivateWork    PublicWork  SelfEmployed    FamilyWork  \
count  11769.000000  11769.000000  11769.000000  11769.000000  11769.000000   
mean    2027.422211     77.434158     16.744549      5.698224      0.124012   
std     1281.005224      7.064767      7.721999      3.482578      0.421047   
min       57.000000     48.200000      0.000000      0.000000      0.000000   
25%     1193.000000     73.400000     11.000000      3.300000      0.000000   
50%     1763.000000     77.800000     16.200000      5.100000      0.000000   
75%     2556.000000     82.300000     21.600000      7.500000      0.000000   
max    12780.000000    100.000000     39.900000     23.300000      3.900000   

       Unemployment  
count  11769.000000  
mean       8.747404  
std        4.728028  
min        0.000000  
25%        5.400000  
50%        7.900000  
75%       11.100000  
max       62.700000  

[8 rows x 38 columns]

In [26]:
print(f'Are there any NaN in the dataset? {blocks_merge.isnull().values.any()}, how many? {blocks_merge.isnull().sum().sum()}')
blocks_merge.info()

Are there any NaN in the dataset? False, how many? 0
<class 'pandas.core.frame.DataFrame'>
Int64Index: 11769 entries, 0 to 11768
Data columns (total 42 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Latitude         11769 non-null  float64
 1   Longitude        11769 non-null  float64
 2   BlockCode        11769 non-null  int64  
 3   County_x         11769 non-null  object 
 4   State            11769 non-null  object 
 5   Tract            11769 non-null  int64  
 6   CensusTract      11769 non-null  int64  
 7   County_y         11769 non-null  object 
 8   Borough          11769 non-null  object 
 9   TotalPop         11769 non-null  int64  
 10  Men              11769 non-null  int64  
 11  Women            11769 non-null  int64  
 12  Hispanic         11769 non-null  float64
 13  White            11769 non-null  float64
 14  Black            11769 non-null  float64
 15  Native           11769 non-null  float64
 16  Asian

In [27]:
blocks_merge.to_csv('blocs_data_census.csv')

#### a result image of all the points from qgis, by borough          

![Image of data blocks](img/nh_blocks.png)